src/uucore/src/lib/features/perms.rs

```diff
@@ -188,6 +189,77 @@ pub struct ChownExecutor {
     pub dereference: bool,
 }
 
+#[cfg(test)]
+pub fn check_root(path: &Path, would_recurse_symlink: bool) -> bool {
+    is_root(path, would_recurse_symlink)
+}
+
+fn is_root(path: &Path, would_recurse_symlink: bool) -> bool {
+    // Prohibit only if:
+    // (--preserve-root and -R present) &&
+    // (path canonicalizes to "/") &&
+    // (
+    //     (path is a symlink && would recurse on this symlink) ||
+    //     (path is not a symlink)
+    // )
+    // The first clause is checked by the caller, the second and third clause is checked here.
+    // The caller has to evaluate -P/-H/-L into 'would_recurse_symlink'.
+    // Recall that canonicalization resolves both relative paths (e.g. "..") and symlinks.
+
+    // The third clause can be evaluated without any syscalls, so we do that first.
+    // If we would_recurse_symlink, then the clause is true no matter whether the path is a symlink
+    // or not. Otherwise, we only need to check here if the path can syntactically be a symlink:
+    if !would_recurse_symlink {
+        // We cannot check path.is_dir() here, as this would resolve symlinks,
+        // which we need to avoid here.
+        // All directory-ish paths match "*/", except ".", "..", "*/.", and "*/..".
+        let looks_like_dir = match path.as_os_str().to_str() {
+            // If it contains special character, prefer to err on the side of safety, i.e. forbidding the chown operation:
+            None => false,
+            Some(".") | Some("..") => true,
+            Some(path_str) => {
+                (path_str.ends_with(MAIN_SEPARATOR_STR))
+                    || (path_str.ends_with(&format!("{}.", MAIN_SEPARATOR_STR)))
+                    || (path_str.ends_with(&format!("{}..", MAIN_SEPARATOR_STR)))
```

Comment:
How would this work when the separator is escaped for instance?

Reply:
I'm not sure what you mean by "escaped", since this should be the already-parsed path. If we're on unix and the string contains some backslash for example, it is an invalid path, because backslashes cannot contain backslashes on unix. If we're on Windows, the semantics of `--preserve-root` go out of the window anyway (pun not intended). Did I miss a potentially-problematic case?

Reply:
I guess not, whenever we do string manipulation like this it usually fails on some escape sequence, but you might be right that there are none here.

Reply:
Ah, I see what you mean. Yes, I'm also uncomfortable with doing this amount of "string-based file system logic". I *think* I didn't miss anything. On the other hand, this is a safety mechanism to prevent bone-headed command-line typos, not a security measure that has to be absolutely waterproof, so a "small" bug in this decision logic shouldn't be a catastrophe.